test/test_helper.rb

```diff
@@ -92,11 +92,31 @@ def with_new_cache
     ViewComponent::CompileCache.cache = Set.new
     old_cache_template_loading = ActionView::Base.cache_template_loading
     ActionView::Base.cache_template_loading = false
+    reset_render_template_methods
 
     yield
   ensure
     ActionView::Base.cache_template_loading = old_cache_template_loading
     ViewComponent::CompileCache.cache = old_cache
+    reset_render_template_methods
+  end
+end
+
+def reset_render_template_methods
+  ViewComponent::Base.descendants.each do |klass|
+    reset_render_template_method(klass)
+  end
+end
+
+def reset_render_template_method(klass)
+  if klass.instance_methods(false).include?(:render_template_for)
+    klass.send(:remove_method, :render_template_for)
+    klass.class_eval <<-RUBY, __FILE__, __LINE__ + 1
+      def render_template_for(variant = nil)
+        self.class.compile(raise_errors: true, force: true)
+        render_template_for(variant)
+      end
+    RUBY
```

Comment:
Unfortunately no. I thought so too, but if I take this code out, tests fail with a `SystemStackError`. The `reset_render_template_method` helper is called by `with_new_cache`, which swaps out the existing cache for a new one, then restores the old one after the test is finished. Now that it's a new day, I can take another look.

Reply:
Alright, so I dug a bit more into this. The problem comes from the fact that invalidating the cache and redefining `#render_template_for` are two different things. When `.with_new_cache` swaps the new cache out for the old one after the `yield`, any `#render_template_for` method that was defined during the execution of the block needs to be reset because it "belongs" to the new, temporary cache. We don't save the old `#render_template_for` method that belonged to the old cache, so we need to ask the compiler to redefine it. Crucially however, _the old cache still contains all the compiled classes_, so the compiler doesn't do anything when the stub `#render_template_for` method in `ViewComponent::Base` is invoked. That's how we end up with infinite recursion, and why I passed `force: true` when redefining the stub method. We could solve this problem by invalidating the old cache, but that kind of defeats the point of swapping caches in and out.

The solution I eventually came up with passes `force: true` to `.compile` in the original `#render_template_for` stub, since we should never allow infinite recursion if we can help it. In all but the test environment, it shouldn't matter anyway.